### PR TITLE
Add HARVEST step to SKIM WF for Run3

### PR DESCRIPTION
**Pull request by youyingli**
#### PR description:

Move SKIM process to step3 from step4 in Run3 skim WF and add HARVEST step to step4 i.e. modify

from 
```
step3 : -s RAW2DIGI,L1Reco,RECO,PAT,NANO,DQM
step4 : -s SKIM
```

to
```
step3 : -s RAW2DIGI,L1Reco,RECO,SKIM,PAT,NANO,DQM
step4 : -s HARVESTING
```
The PR tries to fix the HARVEST missing mentioned in https://github.com/cms-sw/cmssw/pull/42229#issuecomment-1634814582.
Backport of the PR will be prepared after tests approved.

### CODE REVIEW

**Comment by srimanob** [comment]
test parameters:
- workflows = 141.101,141.102,141.103,141.104,141.105,141.106,141.107,141.108,141.109,141.110,141.111,141.112,141.113,141.114
- relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by youyingli** [comment]
The error happens when running ```WElectron``` skim by

```
step3 : -s RAW2DIGI,L1Reco,RECO,SKIM:WElectron,PAT,NANO,DQM:@standardDQM+@miniAODDQM+@nanoAODDQM
```
However, this never happens when running "standard" and "skim" separately as original skim wf

```
step3 : -s RAW2DIGI,L1Reco,RECO,PAT,NANO,DQM:@standardDQM+@miniAODDQM+@nanoAODDQM
step4 : -s SKIM:WElectron
```

Strange! Any suggestion?

**Comment by missirol** [comment]
My understanding is the following.
 - The issue is related to the fact that `trgMatchGsfElectronProducer` internally uses `L1TGlobalUtilHelper` (or `L1GtUtilsHelper`) via `HLTPrescaleProvider`, and those L1T classes try to automatically discover the L1T products they should consume. When there is ambiguity in this choice, an exception is thrown (#39017 contains a few more details).
 - One way to solve this is to specify the correct input products in the python configuration. I would try [*]. @cms-sw/l1-l2 might have a better suggestion, since these are L1T collections.

[*]
````diff
diff --git a/DPGAnalysis/Skims/python/WElectronSkim_cff.py b/DPGAnalysis/Skims/python/WElectronSkim_cff.py
index 1d31286b028..139c6fb9e48 100644
--- a/DPGAnalysis/Skims/python/WElectronSkim_cff.py
+++ b/DPGAnalysis/Skims/python/WElectronSkim_cff.py
@@ -104,10 +104,21 @@ PassingHLT = cms.EDProducer("trgMatchGsfElectronProducer",
     hltTags = cms.untracked.string( HLTPath ),
     triggerEventTag = cms.untracked.InputTag("hltTriggerSummaryAOD","",HLTProcessName),
     triggerResultsTag = cms.untracked.InputTag("TriggerResults","",HLTProcessName),
-    stageL1Trigger = cms.uint32(1)
+    stageL1Trigger = cms.uint32(1),
+    # Stage-1 L1T inputs
+    l1GtRecordInputTag = cms.InputTag('gtDigis'),
+    l1GtReadoutRecordInputTag = cms.InputTag('gtDigis'),
 )
 from Configuration.Eras.Modifier_stage2L1Trigger_cff import stage2L1Trigger
-stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
+stage2L1Trigger.toModify(PassingHLT,
+    stageL1Trigger = 2,
+    # Stage-2 L1T inputs
+    l1tAlgBlkInputTag = cms.InputTag('gtStage2Digis'),
+    l1tExtBlkInputTag = cms.InputTag('gtStage2Digis'),
+    # remove Stage-1 L1T inputs
+    l1GtRecordInputTag = None,
+    l1GtReadoutRecordInputTag = None,
+)
 
 ##    _____             ____        __ _       _ _   _             
 ##   |_   _|_ _  __ _  |  _ \  ___ / _(_)_ __ (_) |_(_) ___  _ __  
````

**Comment by srimanob** [comment]
@youyingli 
Just to clarify, when we do reprocessing with skimming, we do it in the same cmsDriver as in this PR, or we do it separately? 
Anyway, we can specify the proper input as @missirol suggests. Thx.

**Comment by aloeliger** [comment]
> My understanding is the following.
> 
> * The issue is related to the fact that `trgMatchGsfElectronProducer` internally uses `L1TGlobalUtilHelper` (or `L1GtUtilsHelper`) via `HLTPrescaleProvider`, and those L1T classes try to automatically discover the L1T products they should consume. When there is ambiguity in this choice, an exception is thrown ([Consumes failure in `BPHMonitor` #39017](https://github.com/cms-sw/cmssw/issues/39017) contains a few more details).
> * One way to solve this is to specify the correct input products in the python configuration. I would try [*]. @cms-sw/l1-l2 might have a better suggestion, since these are L1T collections.
> 
> [*]
> 
> ```diff
> diff --git a/DPGAnalysis/Skims/python/WElectronSkim_cff.py b/DPGAnalysis/Skims/python/WElectronSkim_cff.py
> index 1d31286b028..139c6fb9e48 100644
> --- a/DPGAnalysis/Skims/python/WElectronSkim_cff.py
> +++ b/DPGAnalysis/Skims/python/WElectronSkim_cff.py
> @@ -104,10 +104,21 @@ PassingHLT = cms.EDProducer("trgMatchGsfElectronProducer",
>      hltTags = cms.untracked.string( HLTPath ),
>      triggerEventTag = cms.untracked.InputTag("hltTriggerSummaryAOD","",HLTProcessName),
>      triggerResultsTag = cms.untracked.InputTag("TriggerResults","",HLTProcessName),
> -    stageL1Trigger = cms.uint32(1)
> +    stageL1Trigger = cms.uint32(1),
> +    # Stage-1 L1T inputs
> +    l1GtRecordInputTag = cms.InputTag('gtDigis'),
> +    l1GtReadoutRecordInputTag = cms.InputTag('gtDigis'),
>  )
>  from Configuration.Eras.Modifier_stage2L1Trigger_cff import stage2L1Trigger
> -stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
> +stage2L1Trigger.toModify(PassingHLT,
> +    stageL1Trigger = 2,
> +    # Stage-2 L1T inputs
> +    l1tAlgBlkInputTag = cms.InputTag('gtStage2Digis'),
> +    l1tExtBlkInputTag = cms.InputTag('gtStage2Digis'),
> +    # remove Stage-1 L1T inputs
> +    l1GtRecordInputTag = None,
> +    l1GtReadoutRecordInputTag = None,
> +)
>  
>  ##    _____             ____        __ _       _ _   _             
>  ##   |_   _|_ _  __ _  |  _ \  ___ / _(_)_ __ (_) |_(_) ___  _ __  
> ```

If I'm understanding the problem as stated correctly (the presence of `gtDigis` and `gtStage2Digis` is causing automatic discovery of the necessary collection to fail) then I would try this as well, it seems the most straightforward simple solution, i.e. just spell out what to use and where.

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py [hunk 1]
```diff
@@ -104,10 +104,22 @@
...
+    l1GtRecordInputTag = cms.InputTag('gtDigis'),
+    l1GtReadoutRecordInputTag = cms.InputTag('gtDigis'),
 )
 from Configuration.Eras.Modifier_stage2L1Trigger_cff import stage2L1Trigger
-stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
+#stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
+stage2L1Trigger.toModify(PassingHLT,
+    stageL1Trigger = 2,
+    # Stage-2 L1T inputs
+    l1tAlgBlkInputTag = cms.InputTag('gtStage2Digis'),
```
Comment (srimanob) [inline DPGAnalysis/Skims/python/WElectronSkim_cff.py:117]: l1tAlgBlkInputTag = 'gtStage2Digis',

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py [hunk 2]
```diff
@@ -104,10 +104,22 @@
...
+    l1tExtBlkInputTag = cms.InputTag('gtStage2Digis'),
```
Comment (srimanob) [inline DPGAnalysis/Skims/python/WElectronSkim_cff.py:118]: Same as above.

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py [hunk 3]
```diff
@@ -104,10 +104,22 @@
...
     triggerResultsTag = cms.untracked.InputTag("TriggerResults","",HLTProcessName),
-    stageL1Trigger = cms.uint32(1)
+    stageL1Trigger = cms.uint32(1),
+    # Stage-1 L1T inputs
+    l1GtRecordInputTag = cms.InputTag('gtDigis'),
+    l1GtReadoutRecordInputTag = cms.InputTag('gtDigis'),
 )
 from Configuration.Eras.Modifier_stage2L1Trigger_cff import stage2L1Trigger
-stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
+#stage2L1Trigger.toModify(PassingHLT, stageL1Trigger = 2)
```
Comment (srimanob) [inline DPGAnalysis/Skims/python/WElectronSkim_cff.py:113]: You don't need to keep the old one. Git already keeps track of modification.

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py (hunk 1 above)
Comment (youyingli) [reply DPGAnalysis/Skims/python/WElectronSkim_cff.py:117]: Isn't ```l1tAlgBlkInputTag``` an InputTag?

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py (hunk 3 above)
Comment (youyingli) [reply DPGAnalysis/Skims/python/WElectronSkim_cff.py:113]: Oh OK, I didn't notice that, thanks! Will remove.

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py (hunk 1 above)
Comment (srimanob) [reply DPGAnalysis/Skims/python/WElectronSkim_cff.py:117]: Are they not defined before? I just look that it passes the existing config and modify. If that is the case, you don't need to redefine the type.

Comment (youyingli) [reply DPGAnalysis/Skims/python/WElectronSkim_cff.py:117]: Yes, they are defined in [here](https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TGlobal/src/L1TGlobalUtilHelper.cc#L10L11). The ```l1tAlgBlkInputTag``` and ```l1tExtBlkInputTag``` will pass through WElectronSKim -> HLTPrescaleProvider -> L1TGlobalUtil-> L1TGlobalUtilHelper.

**Comment by youyingli** [comment]
I used ```runTheMatrix.py -w "standard" -l 141.105``` to test in the local area and it was successful.

File: DPGAnalysis/Skims/python/WElectronSkim_cff.py (hunk 1 above)
Comment (missirol) [reply DPGAnalysis/Skims/python/WElectronSkim_cff.py:116]: > Are they not defined before?

No [*]. Please keep the patch as suggested.

[*] This is the starting module. It does not have `l1tAlgBlkInputTag` nor `l1tExtBlkInputTag`.
```
PassingHLT = cms.EDProducer("trgMatchGsfElectronProducer",    
    InputProducer = cms.InputTag( ELECTRON_COLL ),                          
    hltTags = cms.untracked.string( HLTPath ),
    triggerEventTag = cms.untracked.InputTag("hltTriggerSummaryAOD","",HLTProcessName),
    triggerResultsTag = cms.untracked.InputTag("TriggerResults","",HLTProcessName),
    stageL1Trigger = cms.uint32(1),
    # Stage-1 L1T inputs
    l1GtRecordInputTag = cms.InputTag('gtDigis'),
    l1GtReadoutRecordInputTag = cms.InputTag('gtDigis'),
)
```

**Comment by youyingli** [comment]
@srimanob could you please retrigger it?

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by youyingli** [comment]
@srimanob @sunilUIET please consider signing this PR

**Comment by sunilUIET** [comment]
+pdmv

**Comment by srimanob** [comment]
+Upgrade

**Comment by perrotta** [comment]
+1

**Pull request merged**